clients/cobol-lsp-vscode-extension/src/services/subroutines/SubroutinesCompletionsProvider.ts

```diff
@@ -0,0 +1,87 @@
+import {
+  CancellationToken,
+  CompletionContext,
+  CompletionItem,
+  CompletionItemKind,
+  CompletionItemProvider,
+  Position,
+  Progress,
+  ProgressLocation,
+  TextDocument,
+  window,
+} from "vscode";
+import { SUBROUTINE_COMPLETIONS_SORT_PREFIX } from "../../constants";
+import { listLocalSubroutines } from "./LocalSubroutinesService";
+
+export class SubroutinesCompletionsProvider implements CompletionItemProvider {
+  async provideCompletionItems(
+    document: TextDocument,
+    position: Position,
+    _token: CancellationToken,
+    _context: CompletionContext,
+  ): Promise<CompletionItem[]> {
+    const line = document
+      .lineAt(position.line)
+      .text.slice(0, position.character)
+      .trim();
+
+    const CALL_PATTERN = new RegExp(`^.*\\bCALL(?:\\s+([^\\s]+))?$`, "i");
+
+    const match = line.match(CALL_PATTERN);
+    if (!match) {
+      return [];
+    }
+
+    const prefix = match[1]?.replace(`"`, "").replace(`'`, "");
+
+    let subroutines: string[] = [];
+    let subroutinesLoadingFinished = false;
+    const subroutinesLoadingPromise = (async () => {
+      subroutines = await listLocalSubroutines();
+      subroutinesLoadingFinished = true;
+    })();
+    await Promise.race([
+      subroutinesLoadingPromise,
+      new Promise((resolve) => {
+        setTimeout(resolve, 200);
+      }),
+    ]);
+
+    // If we are unable to provide completions immediately, show info dialog
+    if (!subroutinesLoadingFinished) {
+      await window.withProgress(
```

Comment:
Same question as in the completion case - will this dialog be potentially shown forever?